File: src/content/3.4/code/scala/snippet05.scala
implicit def writerMonad[W: Monoid] = new Functor[Writer[W, ?]] {
  def >=>[A, B, C](f: A => Writer[W, B], g: B => Writer[W, C]) =
    a => {
      val Writer((b, s1)) = f(a)
      val Writer((c, s2)) = g(b)
      Writer((c, Monoid[W].combine(s1, s2)))
    }

  def pure[A](a: A) =
    Writer(a, Monoid[W].empty)
}

object kleisliSyntax {
  //allows us to use >=> as an infix operator
  implicit class MonadOps[M[_], A, B](m1: A => M[B]) {
    def >=>[C](m2: B => M[C])(implicit m: Monad[M]): A => M[C] = {
      m.>=>(m1, m2)
    }
  }
}

File: src/content/3.4/code/ocaml/snippet05.ml
module WriterMonad (W : Monoid) : Monad
with type 'a t = (W.t, 'a) writer =
struct
  type 'a t = (W.t, 'a) writer

  let ( >=> ) f g a =
    let (Writer (b, w)) = f a in
    let (Writer (c, w')) = g b in
    Writer (c, W.mappend w w')

  let return a = Writer (a, W.mempty)
end
